### ISSUE TITLE
HTTP Range header not supported when trying to download an update

### ISSUE THREAD

**Issue by kg** [itchio/itch#2374]
![image](https://user-images.githubusercontent.com/198130/76901189-fe882480-6857-11ea-9f39-8f2656efb20c.png)

Don't see any way to get the actual text of this error message, so here's a screenshot. It doesn't seem to be a transient error, each time I manually check for an update on this game I get the same error a few seconds later.

itch v23.4.2 on Windows 10 x64

**Comment by fasterthanlime** [comment]
Sorry for the craptastic error message.

tl;dr - upgrading to v25 might help, but in the long run, the fix is to contact the developer of the games and tell them to upload *straight to itch.io*, not to plop a github link in the "external upload" box. This can be automated with https://itch.io/docs/butler/.

---

Here's the underlying cause:

  * "External uploads" are a thing on itch.io - I personally wish they weren't a thing, ever, because third-party HTTP servers regularly lack features or just collapse under the traffic thrown at them (no, Apache on your DigitalOcean droplet will not sustain even mild exposure)
  * Back in v23 days, I was ambitious and tried to support external uploads anyway. For stock Apache/nginx, this was relatively easy - provided they're just serving static files, not doing reverse proxy stuff.
  * For GitHub, this was kinda harder. GET'ing a github release redirects an Amazon S3 bucket, and Go's default http client doesn't copy request headers when doing the second request. So, I fixed that, and it started working.

So, maybe something changed in the way GitHub serves downloads (this would make sense after the MS acquisition? Maybe they're moving off of S3 and onto Azure?), or maybe this is not a link to a *release* zip but just a dynamically-generated-on-the-fly tarball/zip file in which case they probably wouldn't support range requests anyway.

I want to say that upgrading to itch v25 would help (grab it from https://itch.io/app - no automatic updates, we changed self-update systems). Other things changed between those two versions, but v25 should find and import your installed games automatically (could take a little while depending on the size of your library).

One of the things in v25 I tried to do is fall back to "just downloading the file to disk first" if something mildly bad happen. Encountering a .rar / .7z archive counts as something mildly bad, as they have no "central directory" (no index = no way to list files = no way to show progress/resume in the middle, etc.). Other worse things can happen, like someone slapping a youtube link in the external upload field, see [this cursed list](https://github.com/itchio/butler/blob/15dca080772b225935f11d28aa0c5d6333dac115/cmd/operate/bad_external_hosts.go).

Before I formally recommended v25 to you I checked with a GitHub release link, and yep, it forces a local download - without range requests - before doing anything else. I'm pretty bummed about the whole thing but... at least that way works.

![image](https://user-images.githubusercontent.com/7998310/76905226-cf96a080-68a1-11ea-8cf6-340e6b6a174e.png)

Anyway. This all sucks. I wish external uploads were never a thing but 🤷

**Comment by kg** [comment]
That makes perfect sense, I forgot about external downloads. It'd be cool if itch v23 was telling me that I needed to install 25 by hand but I'm guessing that's part of why you had to change update systems :)
